core/src/mast/serialization/mod.rs

```diff
@@ -232,35 +229,105 @@ impl Deserializable for MastForest {
         for (node_id, decorator_ids) in before_enter_decorators {
-            let node_id: u32 = node_id.try_into().map_err(|_| {
-                DeserializationError::InvalidValue(format!(
-                    "Invalid node id '{node_id}' while deserializing"
-                ))
-            })?;
-            let node_id = MastNodeId::from_u32_safe(node_id, &mast_forest)?;
+            let node_id = MastNodeId::try_from((node_id, &mast_forest))?;
             mast_forest.set_before_enter(node_id, decorator_ids);
         }
 
         let after_exit_decorators: Vec<(usize, Vec<DecoratorId>)> =
             read_before_after_decorators(source, &mast_forest)?;
         for (node_id, decorator_ids) in after_exit_decorators {
-            let node_id: u32 = node_id.try_into().map_err(|_| {
-                DeserializationError::InvalidValue(format!(
-                    "Invalid node id '{node_id}' while deserializing"
-                ))
-            })?;
-            let node_id = MastNodeId::from_u32_safe(node_id, &mast_forest)?;
+            let node_id = MastNodeId::try_from((node_id, &mast_forest))?;
             mast_forest.set_after_exit(node_id, decorator_ids);
         }
 
         Ok(mast_forest)
     }
 }
 
+fn read_and_validate_magic<R: ByteReader>(source: &mut R) -> Result<[u8; 5], DeserializationError> {
+    let magic: [u8; 5] = source.read_array()?;
+    if magic != *MAGIC {
+        return Err(DeserializationError::InvalidValue(format!(
+            "Invalid magic bytes. Expected '{:?}', got '{:?}'",
+            *MAGIC, magic
+        )));
+    }
+    Ok(magic)
+}
+
+fn read_and_validate_version<R: ByteReader>(
+    source: &mut R,
+) -> Result<[u8; 3], DeserializationError> {
+    let version: [u8; 3] = source.read_array()?;
+    if version != VERSION {
+        return Err(DeserializationError::InvalidValue(format!(
+            "Unsupported version. Got '{version:?}', but only '{VERSION:?}' is supported",
+        )));
+    }
+    Ok(version)
+}
+
+fn read_block_decorators<R: ByteReader>(
+    source: &mut R,
+) -> Result<Vec<(usize, DecoratorList)>, DeserializationError> {
+    let vec_len: usize = source.read()?;
+    let mut out_vec: Vec<_> = Vec::with_capacity(vec_len);
+
+    for _ in 0..vec_len {
+        let node_id: usize = source.read()?;
+
+        let decorator_vec_len: usize = source.read()?;
+        let mut inner_vec: Vec<(usize, DecoratorId)> = Vec::with_capacity(decorator_vec_len);
+        for _ in 0..decorator_vec_len {
+            let op_id: usize = source.read()?;
+            let decorator_id: DecoratorId = source.read()?;
+            inner_vec.push((op_id, decorator_id));
+        }
+
+        out_vec.push((node_id, inner_vec));
+    }
+
+    Ok(out_vec)
+}
+
+fn read_decorator_infos<R: ByteReader>(
+    source: &mut R,
+    decorator_count: usize,
+) -> Result<Vec<DecoratorInfo>, DeserializationError> {
+    let mut decorator_infos: Vec<DecoratorInfo> = Vec::new();
```

Comment:
Ideally we wouldn't need to read this to a `Vec` first. Where this is called, we could iterate over an iterator that reads the next `DecoratorInfo` from `source` and returns it. But I think this requires a custom iterator which is probably overkill.

As an alternative optimization, let's use `Vec::with_capacity(decorator_count)` here so we only allocate once.

Reply:
Decided to implement the iterator since it isn't that much more code. Also implemented a iterator for reading `MastNodeInfo`s